### DISCUSSION TITLE
How to add aquifer in GEOSX?

### DISCUSSION THREAD

**Discussion by wdmhouston** [GEOS-DEV/GEOS#1431] [Q&A] [answered]
Hi Experts,

     Does GEOSX support aquifer in current release? If not,  how to simulate large water body in GEOSX? If yes, can you share an example?

Thanks and Regards,

wdmhouston

**Comment by francoishamon** [comment, marked as answer]
Hello,

Currently we do not support yet analytical aquifers in GEOSX (Carter-Tracy, Fetkovich), but this is a capability that could definitely be very useful for CO2 injection simulations, and that we plan to implement as a special type of face boundary condition in the compositional multiphase solver some time soon. 

However, you can still simulate a numerical aquifer in GEOSX using the multi-region capability. You essentially add (layers of) cells on the sides of the domain with a large pore-volume to mimic the aquifer. You can define define specific fluid properties in the regions corresponding to the aquifer. A setup that we have used in the `InternalMeshGenerator` is for example:
```
     elementTypes="{ C3D8 }"                                                                                                                                                            
      xCoords="{-10000, 0, 32156.4, 42156.4 }"                                                                                                                                           
      yCoords="{-10000, 0, 32156.4, 42156.4 }"                                                                                                                                           
      zCoords="{ 6.096, 91.44 }"                                                                                                                                                         
      nx="{ 10, 64, 10 }"                                                                                                                                                                
      ny="{ 10, 64, 10 }"                                                                                                                                                                
      nz="{ 7 }" 
```
in the `Mesh` block. Then you can use the `cellBlockNames` to define a `CellBlock` in each region. And then in the `ElementRegions` block, for each cell block, you can define the properties of the region:
```
<CellElementRegion                                                                                                                                                                   
      name="aquiferRegion"                                                                                                                                                                     
      cellBlocks="{ cbAquifer }"                                                                                                                                                               
      materialList="{ aquiferFluid, aquiferRock, aquiferRelperm }"/> 
```
If you have issues with the setup of the case, don't hesitate to post your xml file.

**Reply by wdmhouston** [reply]
Great! I'll try.

**Comment by wdmhouston** [comment]
I added an aquifer to Mesh section,
```
  <!-- SPHINX_TUT_INT_HEX_SOLVERS_END -->
  <!-- SPHINX_TUT_INT_HEX_MESH -->
  <Mesh>
<!--main reservoir, usually it's a real geological model and imported as .GRDECL file-->
    <InternalMesh
      name="mesh1"
      elementTypes="{ C3D8 }"
      xCoords="{ 0, 10 }"
      yCoords="{ 0, 10 }"
      zCoords="{ 0, 10 }"
      nx="{ 10 }"
      ny="{ 10 }"
      nz="{ 10 }"
      cellBlockNames="{ cb1 }"/>

<!-- aquifer attached to left-->	  
	<InternalMesh
      name="meshleft"
      elementTypes="{ C3D8 }"
      xCoords="{ -10, 0 }"
      yCoords="{ -10, 0 }"
      zCoords="{ 0, 10 }"
      nx="{ 10 }"
      ny="{ 10 }"
      nz="{ 10 }"
      cellBlockNames="{ cbleft }"/>
  </Mesh>
```
and GEOSX reports error:
```
user@ubuntu:~/aquifer$ geosx -i base_aqifer.xml
Adding Solver of type SinglePhaseFVM, named SinglePhaseFlow
Adding Mesh: InternalMesh, mesh1
Adding Mesh: InternalMesh, meshleft
Adding Geometric Object: Box, source
Adding Geometric Object: Box, sink
Adding Event: PeriodicEvent, solverApplications
Adding Event: PeriodicEvent, outputs
Adding Output: VTK, vtkOutput_aquifer
Adding Object CellElementRegion named mainRegion from ObjectManager::Catalog.
Adding Object CellElementRegion named mainRegion from ObjectManager::Catalog.

***** LOCATION: /apps/GEOSX/src/coreComponents/mainInterface/ProblemManager.cpp:544
***** Controlling expression (should be false): meshBodies.numSubGroups() != 1
***** Rank 0:
Expected meshBodies.numSubGroups() == 1
  meshBodies.numSubGroups() = 2
  1 = 1


** StackTrace of 4 frames **
Frame 0: geosx::ProblemManager::problemSetup()
Frame 1: geosx::GeosxState::initializeDataRepository()
Frame 2: main
Frame 3: __libc_start_main
Frame 4: _start
=====
```
Is it because we can have one InternalMeshGenerator in Mesh section?

**Reply by francoishamon** [reply]
You need to define only one `Mesh`, with one `InternalMesh`, with multiple cell block names in the field `cellBlockNames`. 

You can look at the following integrated tests for complete examples:
https://github.com/GEOSX/GEOSX/blob/68f86fd76b7d2d1caedd176362731764601298f6/src/coreComponents/physicsSolvers/fluidFlow/integratedTests/singlePhaseFlow/compressible_2d_2fluids.xml#L23-L34
 https://github.com/GEOSX/GEOSX/blob/68f86fd76b7d2d1caedd176362731764601298f6/src/coreComponents/physicsSolvers/fluidFlow/integratedTests/singlePhaseFlow/staircase_3d.xml#L23-L34

**Reply by wdmhouston** [reply]
Thanks! I was able to make it work using mesh generator.

**Comment by francoishamon** [comment]
@wdmhouston Regarding your question: _can we use pore volume multiplier to simulate aquifer and how?_

Here is a workaround that you can use for now:

- If your mesh is generated internally in GEOSX, you can provide a table to define the porosity, and nothing prevents you from putting values about one. In the `FieldSpecifications` block, it can look like:
```
 <FieldSpecification                                                                                                                                                                  
      name="ReferencePorosity1"                                                                                                                                                          
      initialCondition="1"                                                                                                                                                               
      setNames="{all}"                                                                                                                                                                   
      objectPath="ElementRegions/reservoir"                                                                                                                              
      fieldName="referencePorosity"                                                                                                                                                      
      scale="1"                                                                                                                                                                          
      functionName="porosityFunction"/> 
``` 
where the `porosityFunction` is defined in the `Functions` block:
```
    <TableFunction                                                                                                                                                                       
      name="porosityFunction"                                                                                                                                                            
      inputVarNames="{elementCenter}"                                                                                                                                                    
      coordinateFiles="{xLinPoro.geos,yLinPoro.geos,zLinPoro.geos}"                                                                                                                      
      voxelFile="poro.geos"                                                                                                                                                              
      interpolation="nearest" />   
```
In `poro.geos`, you can put values above one (simulating a pore volume multiplier) to get cells with very large volume in the aquifer.

An example of how to use a table in GEOSX can be found [here](https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/multiphysics/integratedTests/FieldCaseTutorial1.xml).

- If your mesh is read by PAMELA (GRDECL), you can use a similar workaround with the PORO keyword.

**Comment by wdmhouston** [comment]
Francois,

    That's a good idea!  I will try this!


Regards,

Deming

On Wed, Jun 16, 2021 at 4:08 PM Francois Hamon ***@***.***>
wrote:

> @wdmhouston <https://github.com/wdmhouston> Regarding your question: *can
> we use pore volume multiplier to simulate aquifer and how?*
>
> Here is a workaround that you can use for now:
>
>    - If your mesh is generated internally in GEOSX, you can provide a
>    table to define the porosity, and nothing prevents you from putting values
>    about one. In the FieldSpecifications block, it can look like:
>
>  <FieldSpecification
>       name="ReferencePorosity1"
>       initialCondition="1"
>       setNames="{all}"
>       objectPath="ElementRegions/reservoir"
>       fieldName="referencePorosity"
>       scale="1"
>       functionName="porosityFunction"/>
>
> where the porosityFunction is defined in the Functions block:
>
>     <TableFunction
>       name="porosityFunction"
>       inputVarNames="{elementCenter}"
>       coordinateFiles="{xLinPoro.geos,yLinPoro.geos,zLinPoro.geos}"
>       voxelFile="poro.geos"
>       interpolation="nearest" />
>
> In poro.geos, you can put values above one (simulating a pore volume
> multiplier) to get cells with very large volume in the aquifer.
>
> An example of how to use a table in GEOSX can be found here
> <https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/multiphysics/integratedTests/FieldCaseTutorial1.xml>
> .
>
>    - If your mesh is read by PAMELA (GRDECL), you can use a similar
>    workaround with the PORO keyword.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GEOSX/GEOSX/discussions/1431#discussioncomment-880397>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQFWDEIPKWYFHCMILC5O3TTEHFPANCNFSM45FI6P6Q>
> .
>


-- 
Deming Wang
***@***.***

**Reply by francoishamon** [reply]
@wdmhouston The Carter-Tracy analytical aquifer is being validated right now in #1561, it should be ready to use soon (but not fully matching reference implementations yet)

**Comment by wdmhouston** [comment]
Great! will check!

On Wed, Sep 15, 2021 at 6:49 PM Francois Hamon ***@***.***>
wrote:

> @wdmhouston <https://github.com/wdmhouston> The Carter-Tracy analytical
> aquifer is being validated right now in #1561
> <https://github.com/GEOSX/GEOSX/pull/1561>, it should be ready to use
> soon (but not fully matching reference implementations yet)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GEOSX/GEOSX/discussions/1431#discussioncomment-1335933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQFWFKTJEXEGRDGWINEQTUCEWKLANCNFSM45FI6P6Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Deming Wang
***@***.***

**Reply by francoishamon** [reply]
Hi @wdmhouston

The analytical aquifer is ready to be used, up to date with `develop`, but not yet merged in `develop`. To use it:

- Checkout `feature/hamon/aquifer` and recompile.
- Then, you can try to micmic the setup of the example attached (validated against IX):

1. In the `Geometry` block of the XML file, use `Box` to select the faces that connected the aquifer to the reservoir. This is seems to what `AQUCON` does in the `GRID` section.
```
    <Box
      name="aquifer"
      xMin="{ 999.99, 199.99, 3.99 }"
      xMax="{ 1010.01, 201.01, 6.01 }"/>
```
For now in GEOSX, you need to specify a small buffer (notice the .01 and .99) to make sure that the face is fully enclosed in the box. If you have the impression that the aquifer does not do anything, the first thing to look at is this `Box` that may not contain anything. If this becomes problematic, I will add a `logLevel` to the aquifer.

2. In the `FieldSpecification` block of the XML file, use `Aquifer` to setup the aquifer properties. The input looks like that:
```
   <Aquifer
      name="aquiferBC"
      aquiferPorosity="2e-1"
      aquiferPermeability="3e-13"
      aquiferInitialPressure="9e6"
      aquiferWaterViscosity="0.00089"
      aquiferWaterDensity="962.81"
      aquiferWaterPhaseComponentFraction="{ 0.0, 1.0 }"
      aquiferWaterPhaseComponentNames="{ co2, water }"
      aquiferTotalCompressibility="1e-10"
      aquiferElevation="0"
      aquiferThickness="18"
      aquiferInnerRadius="2000"
      aquiferAngle="20"
      setNames="{ aquifer }"
      objectPath="faceManager"/>
```
where the fields have the same meaning as in `AQUCT`, with the GEOSX units (meters for length, square meters for permeability, Pascal for pressure, Pascal second for viscosity, kg/m^3 for viscosity, degrees for angle).By default, we use the default table of `AQUCT`, but you can use the field `pressureInfluenceFunctionName` to provide an influence table (as a function of dimensionless time).

If you have issues, let's set up a meeting.

[aquifer.zip](https://github.com/GEOSX/GEOSX/files/7238095/aquifer.zip)

**Reply by wdmhouston** [reply]
@francoishamon 
   Great. I'll try this. Thanks.
